---
title: "TCR analysis final report"
output:
  bookdown::html_document2: default
params:
  inputDir: !r getwd()
  chain: !r NULL
  specie: !r NULL
  sampleInfo: !r NULL
---

# Packages {-}
```{r, warning=F, message=FALSE}
library(bookdown)
library(ggplot2)
```

# Work enviroment {-}
```{r}
# Define TCR chain
chain <- params$chain
```

# MiXCR quality control {.tabset}

## Alignment report
```{r plot1,include= TRUE,  echo = F, message= F}
table1 = read.delim("01_alignment_reports.tsv")
table1 = table1[,1:3]
percentage = round(table1$Successfully.aligned.reads/table1$Total.sequencing.reads*100, 3)
table1 = cbind(table1, percentage)
knitr::kable(table1)
```

```{r plot1-1,include= TRUE,  echo = F, message= F}
knitr::include_graphics(path = path.expand("01_alignment_plot.png"))
```

## Assemble report
```{r plot1-3,include= TRUE,  echo = F, message= F}
table2 = read.delim("01_assemble_reports.tsv")
table2 = table1[,1:3]
# percentage = round(table1$Successfully.aligned.reads/table1$Total.sequencing.reads*100, 3)
# table2 = cbind(table2, percentage)
knitr::kable(table2)
```

```{r plot1-4,include= TRUE,  echo = F, message= F}
knitr::include_graphics(path = path.expand("01_assemble_plot.png"))
```


# {-}


# Clones filtering {.tabset}
Functional clonotypes (referring to unique sequences of VDJRegion(nt) after excluding out-of-frames and sequences containing stop codons) were only included in the analysis if they had a minimal abundance read of 2 read counts and their CDR3s were of 4 amino acids minimal length (Greiff et al., 2015, 2017).

## Summary table TCR filtering
```{r plot2-1,include= TRUE,  echo = F, message= F}
table2 = read.delim("02_filtering-evenness_summarytable_TCR.csv")
table2$shannon = round(table2$shannon, 3)
table2 = table2[,c("sample","TCRclones","min_CDR3aa","filterby_readcount","shannon")]
knitr::kable(table2)
```
## Summary table TRA filtering
```{r plot2-2,include= TRUE,  echo = F, message= F}
table2 = read.delim("02_filtering-evenness_summarytable_TRA.csv")
table2$shannon = round(table2$shannon, 3)
table2 = table2[,c("sample","TRAclones","min_CDR3aa","filterby_readcount","shannon")]
knitr::kable(table2)
```
## Summary table TRB filtering
```{r plot2-3,include= TRUE,  echo = F, message= F}
table2 = read.delim("02_filtering-evenness_summarytable_TRB.csv")
table2$shannon = round(table2$shannon, 3)
table2 = table2[,c("sample","TRBclones","min_CDR3aa","filterby_readcount","shannon")]
knitr::kable(table2)
```


# {-}


# Dataset Overview {.tabset}

## Number of cells
The number of cells counted in the cell isolation step.
```{r plot3-1,include= TRUE,  echo = F, message= F, out.width= "70%", out.height="50%"}
knitr::include_graphics(path = path.expand("03_boxplot1_ncell.png"))
```

## RNA quantity
The RNA amount measured in the RNA extraction step.
```{r plot3-2,include= TRUE,  echo = F, message= F, out.width= "70%", out.height="50%", fig.cap= "Figure caption for plot 2."}
knitr::include_graphics(path = path.expand("03_boxplot2_rnaquant.png"))
```

## Reads
Here we represent the successfully aligned reads to TCR that MiXCR found.
```{r plot3-3,include= TRUE,  echo = F, message= F, out.width= "70%", out.height="50%", fig.cap= "Figure caption for plot 2."}
knitr::include_graphics(path = path.expand(paste0("03_boxplot3_reads_", chain, ".png")))
```

## Clones
Clones from MiXCR output after filtering by the number of reads (clones with 2 or more reads are kept, thus avoiding singletons). Additionally, MiXCR output contained BCR+TCR clones and we exclude everything that is not TCR.
```{r plot3-4,include= TRUE,  echo = F, message= F, out.width= "70%", out.height="50%", fig.cap= "Figure caption for plot 2."}
knitr::include_graphics(path = path.expand(paste0("03_boxplot4_clones_", chain, ".png")))
```

## Shannon
Here we represent the Shannon Evenness.
```{r plot3-5,include= TRUE,  echo = F, message= F, out.width= "70%", out.height="50%", fig.cap= "Figure caption for plot 2."}
knitr::include_graphics(path = path.expand(paste0("03_boxplot5_shannon_", chain, ".png")))
```


# {-}


# Correlations
To examine whether the sequencing depth allowed adequate clonal coverage, we determined the correlation of the number of reads, the number of clones and SE vs. the number of cells and then between reads and clones number, reads number and SE, and clones and SE. Ideally, we expect the number of clones to be independent from the number of reads.

From here, all the results correspond to the TCR chain selected by the user ("TRA", "TRB" or "TCR").
```{r plot4,include= TRUE,  echo = F, message= F}
knitr::include_graphics(path = path.expand(paste0("04_correlations_", chain, ".png")))
```


# Overlap analysis (clonal convergence)

Jaccard is a measure of similarity, defined as the size of the intersection divided by the size of the union of the sample sets.
```{r plot5,include= TRUE,  echo = F, message= F}
knitr::include_graphics(path = path.expand(paste0("05_overlap_jaccard_", chain, ".png")))
```

# Diversity analysis (clonal expansion)
```{r plot6,include= TRUE,  echo = F, message= F}
knitr::include_graphics(path = path.expand(paste0("06_diversity_heatmap_", chain, ".png")))
```

# K-mer analysis (repertoire similarity)
```{r plot7,include= TRUE,  echo = F, out.width= "70%", out.height="50%", message= F}
knitr::include_graphics(path = path.expand(paste0("07_kmers_heatmap_", chain, ".png")))
```

# Network analysis (clonal architecture)
The degree of a clone is the number of clones that is similar to (Levenshtein distance of 1 = 1 amino acid change apart).
```{r plot8,include= TRUE,  echo = F, out.width= "50%", out.height="50%", message= F}
knitr::include_graphics(path = path.expand(paste0("08_network_heatmap_", chain, ".png")))
```


# Database analysis (antigen specificity {.tabset}

## McPAS - Fisher test (adjusted p.values)
```{r plot9-1,include= TRUE,  echo = F, message= F}
table1 = read.csv(paste0("09_ddbb_mcpas_", chain, ".csv"))
knitr::kable(table1)
```

## VDJdb - Fisher test (adjusted p.values)
```{r plot9-2,include= TRUE,  echo = F, message= F}
table2 = read.csv(paste0("09_ddbb_vdjdb_", chain, ".csv"))
knitr::kable(table2)
```


# {-}
