---
title: "TCR analysis - CD4 T-cells from MHE patients"
author: Teresa Rubio
date: "March, 2021"
output:
  bookdown::html_document2: default
---

```{r, warning=F, message=FALSE}
library(bookdown)
library(ggplot2)
```

```{r}
# Define working directories
workDir <- "~/MHE_TCR/TCRanalysis_bookdown"
setwd(workDir)
```

# MiXCR quality control {.tabset}

## Alignment report
```{r plot1,include= TRUE,  echo = F, message= F, fig.cap= "Figure caption."}
table1 = read.delim(file.path(workDir, "01_alignment_reports.tsv"))
table1 = table1[,1:3]
percentage = round(table1$Successfully.aligned.reads/table1$Total.sequencing.reads*100, 3)
table1 = cbind(table1, percentage)
knitr::kable(table1)
```

```{r plot1-1,include= TRUE,  echo = F, message= F, fig.cap= "Figure caption."}
knitr::include_graphics(path = path.expand(file.path(workDir, "01_alignment_report.png")))
```

## Assemble report
```{r plot1-3,include= TRUE,  echo = F, message= F, fig.cap= "Figure caption."}
table2 = read.delim(file.path(workDir, "01_assemble_reports.tsv"))
table2 = table1[,1:3]
# percentage = round(table1$Successfully.aligned.reads/table1$Total.sequencing.reads*100, 3)
# table2 = cbind(table2, percentage)
knitr::kable(table2)
```

```{r plot1-4,include= TRUE,  echo = F, message= F, fig.cap= "Figure caption."}
knitr::include_graphics(path = path.expand(file.path(workDir, "01_assemble_report.png")))
```


# {-}


# Clones filtering {.tabset}
Functional clonotypes (referring to unique sequences of VDJRegion(nt) after excluding out-of-frames and sequences containing stop codons) were only included in the analysis if they had a minimal abundance read of 2 read counts and their CDR3s were of 4 amino acids minimal length (Greiff et al., 2015, 2017).

## Summary table TCR filtering
```{r plot2-1,include= TRUE,  echo = F, message= F, fig.cap= "Figure caption."}
table2 = read.delim(file.path(workDir, "02_filtering-evenness_summarytable_TCR.csv"))
table2$shannon = round(table2$shannon, 3)
table2 = table2[,c("sample","TCRclones","min_CDR3aa","filterby_readcount","shannon")]
knitr::kable(table2)
```
## Summary table TRA filtering
```{r plot2-2,include= TRUE,  echo = F, message= F, fig.cap= "Figure caption."}
table2 = read.delim(file.path(workDir, "02_filtering-evenness_summarytable_TRA.csv"))
table2$shannon = round(table2$shannon, 3)
table2 = table2[,c("sample","TRAclones","min_CDR3aa","filterby_readcount","shannon")]
knitr::kable(table2)
```
## Summary table TRB filtering
```{r plot2-3,include= TRUE,  echo = F, message= F, fig.cap= "Figure caption."}
table2 = read.delim(file.path(workDir, "02_filtering-evenness_summarytable_TRB.csv"))
table2$shannon = round(table2$shannon, 3)
table2 = table2[,c("sample","TRBclones","min_CDR3aa","filterby_readcount","shannon")]
knitr::kable(table2)
```


# {-}


# Dataset Overview
```{r plot3,include= FALSE,  echo = F, message= F, out.width= "70%", out.height="50%", fig.cap= "Figure caption."}
knitr::include_graphics(path = path.expand(file.path(workDir, "03_boxplot_merged.png")))
```


## Single Plots {.tabset}

### Number of cells
The number of cells were count with the sorter machine when we isolated CD4 cells.
```{r plot3-1,include= TRUE,  echo = F, message= F, out.width= "70%", out.height="50%", fig.cap= "Figure caption."}
knitr::include_graphics(path = path.expand(file.path(workDir, "03_boxplot_ncell.png")))
```

### RNA quality (RIN)
The RNA amount was measured in the RNA straction step.
```{r plot3-2,include= TRUE,  echo = F, message= F, out.width= "70%", out.height="50%", fig.cap= "Figure caption for plot 2."}
knitr::include_graphics(path = path.expand(file.path(workDir, "03_boxplot_rnaqual.png")))
```

### RNA quantity (ng)
The RNA amount was measured in the RNA straction step.
```{r plot3-3,include= TRUE,  echo = F, message= F, out.width= "70%", out.height="50%", fig.cap= "Figure caption for plot 2."}
knitr::include_graphics(path = path.expand(file.path(workDir, "03_boxplot_rnaquant.png")))
```

### Clones
Clones from mixcr output. In a previous step we have filtered clones by the number of reads (we keep those clones with 2 or more reads, thus avoiding singletones). Additionally, mixcr output contained BCR+TCR clones and we exclude everything that is not TCR.
```{r plot3-4,include= TRUE,  echo = F, message= F, out.width= "70%", out.height="50%", fig.cap= "Figure caption for plot 2."}
knitr::include_graphics(path = path.expand(file.path(workDir, "03_boxplot_clones.png")))
```

### Reads
Here we represent the succesfully aligned reads to TCR that mixcr found.
```{r plot3-5,include= TRUE,  echo = F, message= F, out.width= "70%", out.height="50%", fig.cap= "Figure caption for plot 2."}
knitr::include_graphics(path = path.expand(file.path(workDir, "03_boxplot_reads.png")))
```

### Shannon
Here we represent the Shannon Evenness.
```{r plot3-6,include= TRUE,  echo = F, message= F, out.width= "70%", out.height="50%", fig.cap= "Figure caption for plot 2."}
knitr::include_graphics(path = path.expand(file.path(workDir,"03_boxplot_shannon.png")))
```


# {-}


# Correlations
To examine whether the sequencing depth allowed adequate clonal coverage, we determined the correlation of the number of reads, the number of clones and SE vs. the number of cells and then between reads and clones number, reads number and SE, and clones and SE. Ideally, we expect the number of clones to be independent from the number of reads.

From here, we select TRB (T receptor beta chain) for subsequent analysis due we obtained a major number of clones than using TRA (alpha chain).

## TRB
```{r plot4,include= TRUE,  echo = F, message= F, fig.cap= "Figure caption."}
knitr::include_graphics(path = path.expand(file.path(workDir, "04_correlations_TRB.png")))
```


# Diversity (clonal expansion)
```{r plot5,include= TRUE,  echo = F, message= F, fig.cap= "Figure caption."}
knitr::include_graphics(path = path.expand(file.path(workDir, "05_ClonalExpansion.png")))
```


# Overlap (clonal convergence)
##Jaccard
**method = "jaccard"** - measure of similarity, defined as the size of the intersection divided by the size of the union of the sample sets.
```{r plot6,include= TRUE,  echo = F, message= F, fig.cap= "Figure caption."}
knitr::include_graphics(path = path.expand(file.path(workDir, "06_overlap_jaccard.png")))
```


# K-mers (repertoire similarity)
```{r plot7,include= TRUE,  echo = F, out.width= "70%", out.height="50%", message= F}
knitr::include_graphics(path = path.expand(file.path(workDir, "07_kmers_pearson.png")))
```


# Network (clonal architecture)
## Degree distribution
The degree of a clone is the number of clones that is similar to (LD of 1: 1 amino acid change apart).
The majority of the clones have a degree = 0 in all the samples. 
The maximum degree obtained is 5 (but just a few clones).
The patient with the highest number of conections is PC134.
```{r plot8,include= TRUE,  echo = F, out.width= "50%", out.height="50%", message= F}
knitr::include_graphics(path = path.expand(file.path(workDir, "08_network_degree_heatmap.png")))
```


# Public Databases (Fisher test) {.tabset}

## McPAS - Fisher test (pval)
```{r plot9-1,include= TRUE,  echo = F, message= F}
table1 = read.csv(file.path(workDir, "09_ddbb_Fisher_mcpasTRB.csv"))
knitr::kable(table1)
```

## VDJdb
```{r plot9-2,include= TRUE,  echo = F, message= F}
table2 = read.csv(file.path(workDir, "09_ddbb_Fisher_vdjdbTRB.csv"))
knitr::kable(table2)
```


# {-}


# Methods
## Raw Data Processing 
A total number of 48 raw sequencing files (.fastq.gz files, 24 couple of paired-end reads) were obtained from two sequencing runs. Reads obtained from the same sample and sequencing direction were merged. Merged dataset was processed with MiXCR software package (version 3.0.13) including VDJ alignment and clonotyping (Bolotin et al ., 2015). Clonotypes were assembled by VDJRegion and exported as clones.txt files for further analysis. 

## Clonotypes Filtering
Functional clonotypes (referring to unique sequences of VDJRegion(nt) after excluding out-of-frames and sequences containing stop codons) were only included in the analysis if they had a minimal abundance read of 2 read counts and their CDR3s were of 4 amino acids minimal length (Greiff et al., 2015, 2017).

## Shannon Evenness (SE)
Common diversity indices are special cases of Rényi diversity:
$$
H_a = \frac{1}{1-a} log_b \sum_{i = 1}^{n}{f_i^a}
$$
where $a$ is a scale parameter, $f_i$ is the frequency distribution (proportional abundance of species/clones) and $b$ is the base of the logarithm. It is most popular to use natural logarithms, but some argue for base b = 2 (which makes sense, but no real difference).

Hill (1975) suggested to use so-called ‘Hill numbers’ defined as $N_a = exp(H_a)$:
$$
N_a = (\sum_{i = 1}^{n}{f_i^a}) ^ \frac{1}{1-a}
$$
Some Hill numbers are:
* **Richness**: the number of species with $a = 0$ or $exp(H')$ 
* inverse Simpson with a = 2 and 
* $1/ max(p_i)$ with $a = ∞$
* the exponent of Shannon diversity with $a = 1$.

Diversity is not defined for the case alpha = 1. However, we use L'Hospital's rule to fiind that as alpha tends to 1, Diversity tends to the Shannon entropy. Thus, the Shannon entropy is a special case of the Diversity for alpha = 1. Shannon or Shannon–Weaver (or Shannon–Wiener) index is defined as $H' = − \sum_{i} f_i log_b f_i$

Diversity profile (${}^{a}D$) is defined as:
$$
{}^{a}D = SR \times {}^{a}E
$$
where ${}^{a}E$ is the **evenness** $SR$ is the species **richness** ($SR={}^{a=0}D$), that is, the number of unique clones in a repertoire dataset.

Shannon-Evenness is 1 if all clones in a repertoire have the same frequency (an "even" repertoire), or it converges to 0 if very few clones dominate in the repertoire ("polarized" repertoire).

We use the functions diversity(index="shanon") and richness(index="observed") of the R package microbiome to calculate Shannon Evenness.


## Clonal Expansion
Hill evenness profiles have been developed in mathematical ecology and allow a finegrained comparison of clonal expansion across patient repertoire. Here we calculate pairwise Pearson Correlation of the Evenness profiles, defined as $^{a}D / SR$, where $a$ ranges from 0 to 10 with a step size of 0.2. For visualization, we generated a heatmap using the NMF R package (aheatmap function).