---
title: "03 - Dataset Overview"
author: "Teresa Rubio"
date: "February, 2021"
output:
    html_document:
      code_folding: hide
params:
  inputDir1: !r getwd()
  inputDir2: !r getwd()
  workDir: !r getwd()
  outputDir: !r getwd()
  sampleInfo: !r NULL
---

# Packages
```{r, message=FALSE}
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggpubr)
```

# Work enviroment
```{r}
# Define input/work/output directories
sampleInfoFile <- params$sampleInfo
dir.create(params$outputDir, recursive=T)

# Load sample info
SampleInfo <- read.table(sampleInfoFile, sep = ",", header = T, check.names = F)
levels <- unique(SampleInfo$Group)
SampleInfo$Group <- factor(SampleInfo$Group, levels = levels)

# Color palette for sample groups
cbPalette <- c("#56B4E9","#CB181D","#000000","#CC79A7","#db6d00","#009E73",
               "#006ddb","#b66dff","#924900","#490092","#999999","#F0E442")
```

# Dataset Overview
Dataset overview is the bias detection in several repertoire characteristics: number of cells, RNA amount, RNA quality, reads, clones and Shannon evenness.


## Number of cells
The number of cells were count with the sorter machine when we isolated CD4 cells.
We represent boxplots by group and performed Kruskal-Wallis test to detect possible bias.
```{r, warning=FALSE}
my_comparisons <- combn(levels, m = 2, simplify = F)

# Kruskal-Wallis test
p1_ncell <- ggboxplot(SampleInfo,
                      x = "Group",
                      y = "CellNumber",
                      fill = "cadetblue1",
                      alpha = 0.5,
                      show.legend = FALSE,
                      add = "jitter") +
            theme(legend.position = "none",
                  text = element_text(size=20)) +
            stat_compare_means(comparisons = my_comparisons,
                               method = "wilcox.test",
                               size = 7) +
            stat_compare_means(method = "kruskal",
                               label.y = 1500000,
                               size = 7)

# plots
p1_ncell
```

### Final plot
This is the final plot to include in the report:
```{r, results='hide'}
save(p1_ncell, file = file.path(params$workDir, "03_boxplot1_ncell.Rda"))

png(file.path(params$outputDir, "03_boxplot1_ncell.png"), height = 30, width = 30, units = "cm", res = 300)
p1_ncell
dev.off()
```


## RNA amount
The RNA amount was measured in the RNA extraction step.
```{r}
# Kruskal-Wallis test
p1_totalRNA <- ggboxplot(SampleInfo,
                      x = "Group",
                      y = "totalRNA",
                      fill = "blueviolet",
                      alpha = 0.2,
                      add = "jitter") +
            theme(legend.position = "none",
                  text = element_text(size=20)) +
            stat_compare_means(comparisons = my_comparisons,
                               method = "wilcox.test",
                               size = 7) +
            stat_compare_means(method = "kruskal",
                               label.y = 1300,
                               size = 7)

# plots
p1_totalRNA
```

### Final plot
This is the final plot to include in the report:
```{r, results='hide'}
save(p1_totalRNA, file = file.path(params$workDir, "03_boxplot2_rnaquant.Rda"))

png(file.path(params$outputDir, "03_boxplot2_rnaquant.png"), height = 30, width = 30, units = "cm", res = 300)
p1_totalRNA
dev.off()
```


## Reads
### RNA-seq reads
The number of reads in the complete RNA-seq dataset is represented.
```{r}
# Kruskal-Wallis test
p1_reads <- ggboxplot(SampleInfo,
                      x = "Group",
                      y = "NumReads",
                      color = "Group",
                      palette = cbPalette,
                      add = "jitter") +
            theme(legend.position = "none",
                  text = element_text(size=20)) +
            stat_compare_means(comparisons = my_comparisons,
                               method = "wilcox.test") +
            stat_compare_means(method = "kruskal",
                               label.y = 77011995)

# plots
p1_reads
```

### TRA reads
Here we represent the successfully aligned reads to TCA (alpha chain) that mixcr found.
```{r}
alignment_reports <- read.delim("01_alignment_reports.tsv")

# Calculating all TRA reads
TRA_reads <- alignment_reports[,c("SampleID", "TRA.chains")]
colnames(TRA_reads)[2] <- "TRAreads"

# Add one column to SampleInfo
SampleInfo <- merge(SampleInfo, TRA_reads[,c("SampleID", "TRAreads")],
                    by.x = "SampleID", by.y = "SampleID", sort = F)

# Kruskal-Wallis test
p1_readsTRA <- ggboxplot(SampleInfo,
                         x = "Group",
                         y = "TRAreads",
                         fill = "pink",
                         alpha = 0.5,
                         add = "jitter") +
               theme(legend.position = "none",
                     text = element_text(size=20)) +
               stat_compare_means(comparisons = my_comparisons,
                                  method = "wilcox.test") +
               stat_compare_means(method = "kruskal",
                                  label.y = 14000)

# plots
p1_readsTRA
```

### TRB reads
Here we represent the successfully aligned reads to TRB (beta chain) that mixcr found.
```{r}
# Calculating all TRB reads
TRB_reads <- alignment_reports[,c("SampleID", "TRB.chains")]
colnames(TRB_reads)[2] <- "TRBreads"

# Add one column to SampleInfo
SampleInfo <- merge(SampleInfo, TRB_reads[,c("SampleID", "TRBreads")],
                    by.x = "SampleID", by.y = "SampleID", sort = F)

# Kruskal-Wallis test
p1_readsTRB <- ggboxplot(SampleInfo,
                         x = "Group",
                         y = "TRBreads",
                         fill = "pink",
                         alpha = 0.5,
                         add = "jitter") +
                theme(legend.position = "none",
                      text = element_text(size=20)) +
                stat_compare_means(comparisons = my_comparisons,
                                   method = "wilcox.test") +
                stat_compare_means(method = "kruskal",
                                   label.y = 30000)

# plots
p1_readsTRB
```

### Final plot
Final plots of the number of reads in TRA and TRB.
```{r, results='hide'}
# TRA plot
save(p1_readsTRA, file = file.path(params$workDir, "03_boxplot3_reads_TRA.Rda"))
png(file.path(params$outputDir, "03_boxplot3_reads_TRA.png"), height = 30, width = 30, units = "cm", res = 300)
p1_readsTRA
dev.off()

# TRB plot
save(p1_readsTRB, file = file.path(params$workDir, "03_boxplot3_reads_TRB.Rda"))
png(file.path(params$outputDir, "03_boxplot3_reads_TRB.png"), height = 30, width = 30, units = "cm", res = 300)
p1_readsTRB
dev.off()
```


## Clones
Clones from MiXCR assemble report (before filtering).

### TRA clones
```{r}
summary_table_TRA <- read.delim("02_filtering-evenness_summarytable_TRA.csv")
summary_table_TRA <- summary_table_TRA[,c("sample", "filterby_readcount")]
colnames(summary_table_TRA) <- c("sample", "TRAclones")

SampleInfo <- merge(SampleInfo, summary_table_TRA, by.x = "SampleID", by.y = "sample", sort = F)

# Kruskal-Wallis test
p1_clones_filtered_TRA <- ggboxplot(SampleInfo,
                                    x = "Group",
                                    y = "TRAclones",
                                    fill = "chartreuse",
                                    alpha = 0.5,
                                    add = "jitter") +
                          theme(legend.position = "none",
                                text = element_text(size=20)) +
                          stat_compare_means(comparisons = my_comparisons,
                                             method = "wilcox.test") +
                          stat_compare_means(method = "kruskal",
                                             label.y = 1500)
p1_clones_filtered_TRA
```

### TRB clones
```{r}
summary_table_TRB <- read.delim("02_filtering-evenness_summarytable_TRB.csv")
summary_table_TRB <- summary_table_TRB[,c("sample", "filterby_readcount")]
colnames(summary_table_TRB) <- c("sample", "TRBclones")

SampleInfo <- merge(SampleInfo, summary_table_TRB, by.x = "SampleID", by.y = "sample", sort = F)

# Kruskal-Wallis test
p1_clones_filtered_TRB <- ggboxplot(SampleInfo,
                                    x = "Group",
                                    y = "TRBclones",
                                    fill = "chartreuse",
                                    alpha = 0.5,
                                    add = "jitter") +
                          theme(legend.position = "none",
                                text = element_text(size=20)) +
                          stat_compare_means(comparisons = my_comparisons,
                                             method = "wilcox.test") +
                          stat_compare_means(method = "kruskal",
                                             label.y = 1500)
p1_clones_filtered_TRB
```

### Final plot
Final plots of the number of clones in TRA and TRB.
```{r, results='hide'}
# TRA plot
save(p1_clones_filtered_TRA, file = file.path(params$workDir, "03_boxplot4_clones_TRA.Rda"))
png(file.path(params$outputDir, "03_boxplot4_clones_TRA.png"), height = 30, width = 30, units = "cm", res = 300)
p1_clones_filtered_TRA
dev.off()

# TRB plot
save(p1_clones_filtered_TRB, file = file.path(params$workDir, "03_boxplot4_clones_TRB.Rda"))
png(file.path(params$outputDir, "03_boxplot4_clones_TRB.png"), height = 30, width = 30, units = "cm", res = 300)
p1_clones_filtered_TRB
dev.off()
```

## Shannon Evenness
In the filtering step we also calculated Shannon Evenness as defined in (Greiff et al. Genome medicine, 2015, 7:49).

### TRA Shannon-Evenness
```{r}
summary_table_TRA <- read.delim("02_filtering-evenness_summarytable_TRA.csv", check.names = FALSE)
summary_table_TRA <- summary_table_TRA[,c("sample", "Shannon-Evenness")]
summary_table_TRA$`Shannon-Evenness` <- round(summary_table_TRA$`Shannon-Evenness`, 4)
colnames(summary_table_TRA) <- c("sample", "TRAshannonevenness")

SampleInfo <- merge(SampleInfo, summary_table_TRA, by.x = "SampleID", by.y = "sample", sort = F)

# Kruskal-Wallis test
p1_shannon_filtered_TRA <- ggboxplot(SampleInfo,
                                     x = "Group",
                                     y = "TRAshannonevenness",
                                     fill = "yellow",
                                     add = "jitter") +
                           theme(legend.position = "none",
                                 text = element_text(size=20)) +
                           stat_compare_means(comparisons = my_comparisons,
                                              method = "wilcox.test") +
                           stat_compare_means(method = "kruskal",
                                              label.y =  1.3)
p1_shannon_filtered_TRA
```

### TRB Shannon-Evenness
```{r}
summary_table_TRB <- read.delim("02_filtering-evenness_summarytable_TRB.csv", check.names = FALSE)
summary_table_TRB <- summary_table_TRB[,c("sample", "Shannon-Evenness")]
summary_table_TRB$`Shannon-Evenness` <- round(summary_table_TRB$`Shannon-Evenness`, 4)
colnames(summary_table_TRB) <- c("sample", "TRBshannonevenness")

SampleInfo <- merge(SampleInfo, summary_table_TRB, by.x = "SampleID", by.y = "sample", sort = F)

# Kruskal-Wallis test
p1_shannon_filtered_TRB <- ggboxplot(SampleInfo,
                                     x = "Group",
                                     y = "TRBshannonevenness",
                                     fill = "yellow",
                                     add = "jitter") +
                           theme(legend.position = "none",
                                 text = element_text(size=20)) +
                           stat_compare_means(comparisons = my_comparisons,
                                              method = "wilcox.test") +
                           stat_compare_means(method = "kruskal",
                                              label.y = 1.3)
p1_shannon_filtered_TRB
```

### Final plot
Final plots of the Shannon Eveness in TRA and TRB clones.
```{r, results='hide'}
# TRA plot
save(p1_shannon_filtered_TRA, file = file.path(params$workDir, "03_boxplot5_shannonevenness_TRA.Rda"))
png(file.path(params$outputDir, "03_boxplot5_shannonevenness_TRA.png"), height = 30, width = 30, units = "cm", res = 300)
p1_shannon_filtered_TRA
dev.off()

# TRB plot
save(p1_shannon_filtered_TRB, file = file.path(params$workDir, "03_boxplot5_shannonevenness_TRB.Rda"))
png(file.path(params$outputDir, "03_boxplot5_shannonevenness_TRB.png"), height = 30, width = 30, units = "cm", res = 300)
p1_shannon_filtered_TRB
dev.off()
```

# Output
Columns have been added to the SampleInfo file: 
* ReadPairsNumber: total number of reads in the .fastq raw files.
* TRAreads:  succesfully aligned reads to TRA chain 
* TRBreads:  succesfully aligned reads to TRB chain 
* Clones:    total number of assembled clones (B and T-cells) 
* TRAclones: total number of assembled TRA clones
* TRBclones: total number of assembled TRB clones
* TRAshannonevenness: Shannon eveness of TRA clones
* TRBshannonevenness: Shannon eveness of TRB clones
```{r}
write.table(SampleInfo, file = file.path(params$outputDir, "03_datasetoverview.csv"),
            sep = ",", row.names = F, col.names = T, quote = F)
```
