---
title: "TCR analysis final report"
output:
  bookdown::html_document2: default
params:
  inputDir: !r getwd()
  chain: !r NULL
  specie: !r NULL
  sampleInfo: !r NULL
---

```{r, warning=F, message=FALSE, include=FALSE}
library(bookdown)
library(ggplot2)

# Define TCR chain
chain <- params$chain
```

# MiXCR quality control {.tabset}

## Alignment report

* Total sequencing reads: total number of analysed sequencing.
* Successfully aligned reads: number of successful alignments.

```{r plot1,include= TRUE,  echo = F, message= F}
table1 <- read.delim("01_alignment_reports.tsv", check.names = FALSE)
table1 <- table1[,c("SampleID",	"Total sequencing reads",	"Successfully aligned reads")]
Percentage <- round(table1$`Successfully aligned reads`/table1$`Total sequencing reads`*100, 3)
table1 <- cbind(table1, Percentage)
knitr::kable(table1)
```

```{r plot1-1,include= TRUE,  echo = F, message= F}
knitr::include_graphics(path = path.expand("01_alignment_plot.png"))
```

## Assemble report

* Final clonotype count: number of clonotypes after all MiXCR error correction steps.
* Reads used in clonotypes: sum of all clonotype abundances.

```{r plot1-3,include= TRUE,  echo = F, message= F}
table2 <- read.delim("01_assemble_reports.tsv", check.names = FALSE)
table2 <- table2[,c("SampleID",	"Final clonotype count", "Reads used in clonotypes, percent of total")]
colnames(table2)[3] <- "Reads used in clonotypes"
knitr::kable(table2)
```

```{r plot1-4,include= TRUE,  echo = F, message= F}
knitr::include_graphics(path = path.expand("01_assemble_plot.png"))
```


# {-}


# Clone filtering {.tabset}
Functional clonotypes (referring to unique sequences of VDJRegion(nt) after excluding out-of-frames and sequences containing stop codons) were only included in the analysis if they had a minimal abundance read of 2 read counts and their CDR3s were of 4 amino acids minimal length as previously described (R. Amoriello et al., 2020).

## Summary table TRA filtering
```{r plot2-2,include= TRUE,  echo = F, message= F}
table2 <- read.delim("02_filtering-evenness_summarytable_TRA.csv", check.names = FALSE)
table2 <- table2[,c("sample", "TRAclones", "filterby_both", "Shannon-Evenness")]
colnames(table2) <- c("SampleID", "Number of TRA clones", "Number of filtered clones", "Shannon-Evenness")
table2$`Shannon-Evenness` <- round(table2$`Shannon-Evenness`, 3)
knitr::kable(table2)
```
## Summary table TRB filtering
```{r plot2-3,include= TRUE,  echo = F, message= F}
table2 <- read.delim("02_filtering-evenness_summarytable_TRB.csv", check.names = FALSE)
table2 <- table2[,c("sample", "TRBclones", "filterby_both", "Shannon-Evenness")]
colnames(table2) <- c("SampleID", "Number of TRB clones", "Number of filtered clones", "Shannon-Evenness")
table2$`Shannon-Evenness` <- round(table2$`Shannon-Evenness`, 3)
knitr::kable(table2)
```


# {-}


# Dataset overview {.tabset}

## Number of cells
The number of cells counted in the cell isolation step.
```{r plot3-1,include= TRUE,  echo = F, message= F, out.width= "70%", out.height="50%"}
knitr::include_graphics(path = path.expand("03_boxplot1_ncell.png"))
```

## RNA quantity
The RNA amount measured in the RNA extraction step.

```{r plot3-2,include= TRUE,  echo = F, message= F, out.width= "70%", out.height="50%"}
knitr::include_graphics(path = path.expand("03_boxplot2_rnaquant.png"))
```

## Reads
The number of successfully aligned reads to ``r params$chain`` by MiXCR.

```{r plot3-3,include= TRUE,  echo = F, message= F, out.width= "70%", out.height="50%"}
knitr::include_graphics(path = path.expand(paste0("03_boxplot3_reads_", chain, ".png")))
```

## Clones
The number of assembled clones by MiXCR after the filtering step.

```{r plot3-4,include= TRUE,  echo = F, message= F, out.width= "70%", out.height="50%"}
knitr::include_graphics(path = path.expand(paste0("03_boxplot4_clones_", chain, ".png")))
```

## Shannon Evenness (SE)
Shannon Evenness (SE) values per repertoire.

```{r plot3-5,include= TRUE,  echo = F, message= F, out.width= "70%", out.height="50%"}
knitr::include_graphics(path = path.expand(paste0("03_boxplot5_shannonevenness_", chain, ".png")))
```


# {-}


# Correlations
To examine whether the sequencing depth allowed adequate clonal coverage, we determined the correlation of the number of reads, the number of clones and Shannon Evenness (SE) vs. the number of cells and then between reads and clones number, reads number and SE, and clones and SE. Ideally, we expect the number of clones to be independent from the number of reads.

NOTE: From here, all the results correspond to the TCR chain selected by the user = ``r params$chain``.
```{r plot4,include= TRUE,  echo = F, message= F}
knitr::include_graphics(path = path.expand(paste0("04_correlations_", chain, ".png")))
```


# Overlap analysis (clonal convergence)
Repertoire overlap or clonal convergence analysis is the most common approach to uncover clonotypes shared between given individuals, which are also denominated as “public” clones.
Pairwise clonal convergence between two repertoires A and B was quantified using the Jaccard similarity coefficient, defined as the size of the intersection divided by the size of the union of the sample sets.
```{r plot5,include= TRUE,  echo = F, message= F}
knitr::include_graphics(path = path.expand(paste0("05_overlap_jaccard_", chain, ".png")))
```

# Diversity analysis (clonal expansion)
The expansion of individual T-cell clones that bind their matching antigen can be analyzed using Hill-based evenness profiles, a diversity measurement derived from ecology. Unlike single diversity indices, which can produce different clonal expansion results, diversity profiles capture the entire immune repertoire and reflect immunological statuses more sensitively (V. Greiff et al., 2015).
```{r plot6,include= TRUE,  echo = F, message= F}
knitr::include_graphics(path = path.expand(paste0("06_diversity_heatmap_", chain, ".png")))
```

# K-mer analysis (repertoire similarity)
The adaptive immune response is determined by immune receptor sequences: the higher their dissimilarity, the wider the range of antigens they are able to recognize. The all-to-all sequence similarity within a repertoire represents the repertoire architecture, which was measured in the samples using both k-mers and network analysis.
Overlapping k-mers of length 3 (k = 3) were extracted from the amino acid CDR3 sequences in each TCR repertoire and condensed into a k-mer frequency distribution as previously described in R. Amoriello et al., 2020. Hierarchical clustering was performed based on euclidean distance for correlations and heatmaps were generated for visualization.
```{r plot7,include= TRUE,  echo = F, out.width= "70%", out.height="50%", message= F}
knitr::include_graphics(path = path.expand(paste0("07_kmers_heatmap_", chain, ".png")))
```

# Network analysis (clonal architecture)
Repertoire similarity networks were generated as previously described (R. Amoriello et al., 2020), where nodes represent amino acid CDR3 sequences and edges were drawn between sequences differing by 1 amino acid (Levenshtein distance = 1). The degree (number of linked nodes) distributions of each repertoire were calculated and heatmaps were generated for visualization.
```{r plot8,include= TRUE,  echo = F, out.width= "50%", out.height="50%", message= F}
knitr::include_graphics(path = path.expand(paste0("08_network_heatmap_", chain, ".png")))
```


# Database analysis (antigen specificity) {.tabset}
The overrepresentation of clones associated with diseases or antigens in the CDR3 sequences was evaluated by a one-tailed Fisher’s exact test applied to each group of samples, using the disease categories included in the McPAS-TCR and VDJdb databases.

## McPAS - Fisher test (adjusted p.values)
```{r plot9-1,include= TRUE,  echo = F, message= F}
table1 <- read.csv(paste0("09_ddbb_mcpas_", chain, ".csv"))
knitr::kable(table1)
```

## VDJdb - Fisher test (adjusted p.values)
```{r plot9-2,include= TRUE,  echo = F, message= F}
table2 <- read.csv(paste0("09_ddbb_vdjdb_", chain, ".csv"))
knitr::kable(table2)
```

# {-}

# Methods
More information about the applied methods can be found in the paper called "*A Nextflow pipeline for CD4 T-cell repertoire reconstruction and analysis from RNA sequencing data*" by T. Rubio, M. Chernigovskaya, C. Marti, P. Izquierdo-Altarejos, A. Urios, C. Montoliu, V. Felipo, A. Conesa, V. Greiff and S. Tarazona. 

# References
* R. Amoriello, V. Greiff, A. Aldinucci, E. Bonechi, A. Carnasciali, B. Peruzzi, A.M. Repice, A. Mariottini, R. Saccardi, B. Mazzanti, L. Massacesi, C. Ballerini, *The TCR Repertoire Reconstitution in Multiple Sclerosis: Comparing One-Shot and Continuous Immunosuppressive Therapies*, Front. Immunol. 11 (2020) 559. https://doi.org/10.3389/fimmu.2020.00559.
* V. Greiff, P. Bhat, S.C. Cook, U. Menzel, W. Kang, S.T. Reddy, *A bioinformatic framework for immune repertoire diversity profiling enables detection of immunological status*, Genome Med. 7 (2015) 49. https://doi.org/10.1186/s13073-015-0169-8.

